component/src/stake/component.rs

```diff
@@ -911,6 +918,39 @@ pub trait StateReadExt: StateRead {
             .unwrap_or_default()
     }
 
+    async fn penalty_in_epoch(
+        &self,
+        id: &IdentityKey,
+        epoch_index: u64,
+    ) -> Result<Option<Penalty>> {
+        self.get(&state_key::penalty_in_epoch(id, epoch_index))
+            .await
+    }
+
+    /// Returns the compounded penalty for the given validator over the half-open range of epochs [start, end).
+    async fn compounded_penalty_over_range(
+        &self,
+        id: &IdentityKey,
+        start: u64,
+        end: u64,
+    ) -> Result<Penalty> {
+        let prefix = state_key::penalty_in_epoch_prefix(id);
+        let all_penalties = self
+            .prefix::<Penalty, _>(&prefix)
+            .try_collect::<BTreeMap<String, Penalty>>()
+            .await?;
+
+        let start_key = state_key::penalty_in_epoch(id, start);
+        let end_key = state_key::penalty_in_epoch(id, end);
+
+        let mut compounded = Penalty::default();
+        for (_key, penalty) in all_penalties.range(start_key..end_key) {
+            compounded = compounded.compound(*penalty);
+        }
```

Comment:
This pulls more data than it could, but we only do prefix queries, not range queries, and also efficiency doesn't matter here because there shouldn't be very many slashings.